### PR TITLE
STM32: Add never_reset reservation to RGBMatrix init

### LINKED ISSUE #3440: rgbmatrix: hang when soft resetting
I encountered this on STM32 with a build derived from 6.0.0-alpha.3-267-gd774678a0, but from the traceback it seems like it MIGHT be a general problem.  Or it MIGHT not.

You don't need a matrix attached to reproduce this.

Just run:
```
>>> import board, rgbmatrix, framebufferio; m = rgbmatrix.RGBMatrix( width=64, bit_depth=4, rgb_pins=[board.D13, board.D12, board.D11, board.A4, board.A5, board.D6], addr_pins=[board.A0, board.A1, board.A2, board.A3], clock_pin=board.D5, latch_pin=board.D10, output_enable_pin=board.D9); d = framebufferio.FramebufferDisplay(m)
```
and hit ctrl-d.  It'll hang, usually right after printing "code.py output:".  Debugger says:
```
Program received signal SIGTRAP, Trace/breakpoint trap.
[Switching to Thread 57005]
_PM_swapbuffer_maybe (core=0x2000306c <displays+20>)
    at ../../lib/protomatter/arch.h:1579
1579	      ;
(gdb) where
#0  _PM_swapbuffer_maybe (core=0x2000306c <displays+20>)
    at ../../lib/protomatter/arch.h:1579
#1  0x0803f1c4 in _refresh_display (self=self@entry=0x20003118 <displays+192>)
    at ../../shared-module/framebufferio/FramebufferDisplay.c:271
#2  0x0803f618 in framebufferio_framebufferdisplay_background (
    self=self@entry=0x20003118 <displays+192>)
    at ../../shared-module/framebufferio/FramebufferDisplay.c:348
#3  0x0803e9dc in displayio_background ()
    at ../../shared-module/displayio/__init__.c:71
#4  0x080213da in supervisor_background_tasks (unused=<optimized out>)
    at ../../supervisor/shared/tick.c:80
#5  0x08020642 in background_callback_run_all ()
    at ../../supervisor/shared/background_callback.c:87
#6  0x08021488 in supervisor_run_background_tasks_if_tick ()
    at ../../supervisor/shared/tick.c:141
#7  0x0800f79c in mp_obj_print_helper (print=0x2001fdd0, o_in=0x20019a90, 
    kind=<optimized out>) at ../../py/obj.c:68
#8  0x0801b3fa in mp_builtin_print (n_args=<optimized out>, 
    pos_args=<optimized out>, kw_args=<optimized out>)
    at ../../py/modbuiltins.c:411
#9  0x0801268e in fun_builtin_var_call (args=0x200003f0 <_pystack+24>, n_kw=0, 
    n_args=1, self_in=0x805b5a8 <mp_builtin_print_obj>)
    at ../../py/objfun.c:124
#10 fun_builtin_var_call (self_in=0x805b5a8 <mp_builtin_print_obj>, n_args=1, 
    n_kw=0, args=0x200003f0 <_pystack+24>) at ../../py/objfun.c:110
#11 0x0801bd46 in mp_execute_bytecode (
    code_state=code_state@entry=0x200003d8 <_pystack>, 
    inject_exc=<optimized out>, inject_exc@entry=0x0) at ../../py/vm.c:934
#12 0x0801261e in fun_bc_call (self_in=0x20019a70, n_args=0, n_kw=0, args=0x0)
    at ../../py/objfun.c:286
#13 0x0802bd52 in parse_compile_execute (source=source@entry=0x805cb10, 
    input_kind=input_kind@entry=MP_PARSE_FILE_INPUT, 
    exec_flags=exec_flags@entry=32, result=result@entry=0x2001ffa0)
    at ../../lib/utils/pyexec.c:114
#14 0x0802c106 in pyexec_file (filename=filename@entry=0x805cb10 "code.py", 
    result=result@entry=0x2001ffa0) at ../../lib/utils/pyexec.c:531
#15 0x0801fcaa in maybe_run_list (
    filenames=filenames@entry=0x805cb64 <supported_filenames>, 
    exec_result=exec_result@entry=0x2001ffa0) at ../../main.c:214
#16 0x0801fddc in run_code_py (safe_mode=safe_mode@entry=NO_SAFE_MODE)
    at ../../main.c:285
#17 0x080200be in main () at ../../main.c:519
(gdb) list
1574	  if (core->doubleBuffer) {
1575	    core->swapBuffers = 1;
1576	    // To avoid overwriting data on the matrix, don't return
1577	    // until the timer ISR has performed the swap at the right time.
1578	    while (core->swapBuffers)
1579	      ;
1580	  }
1581	}
1582	
1583	#endif // ARDUINO || CIRCUITPYTHON
```
My interpretation is that we are waiting for the interrupt routine to run.  If this is STM specific there's something about the timer/interrupt state that is not being carried over the soft reset correctly.  I recommend checking against a SAMD51 board for the same bug, and if it's not there then pursue it as an STM timer problem.

### PR DESCRIPTION
**Pull request by hierophect**
When soft-reset on STM32, RGBMatrix currently gets stuck in `_PM_swapbuffer_maybe`, as the timer IRQ it relies on to exit an infinite loop is disabled by the reset process. This PR adds timers used by the RGBMatrix constructor to the never_reset array so this does not happen. Tested on STM32F405 Feather Express.

Also removes some timer debug messages that were accidentally left in from a previous PR.

Resolves #3440. Mentioned in #3051 but ultimately unrelated to the issue there, which seems to have just been a documentation mixup.

### CODE REVIEW

File: ports/stm/common-hal/rgbmatrix/RGBMatrix.c [hunk 1]
```diff
@@ -36,6 +36,7 @@ extern void _PM_IRQ_HANDLER(void);
 void *common_hal_rgbmatrix_timer_allocate() {
     TIM_TypeDef * timer = stm_peripherals_find_timer();
     stm_peripherals_timer_reserve(timer);
+    stm_peripherals_timer_never_reset(timer);
```
Comment (jepler) [inline ports/stm/common-hal/rgbmatrix/RGBMatrix.c:39]: with "never_reset", does `stm_peripherals_timer_free` in `common_hal_rgbmatrix_timer_free()` remain enough to fully deallocate/free/reset the timer?

**Review by jepler — APPROVED** [review]
If the answer to my question is "yes", this is good to merge.

File: ports/stm/common-hal/rgbmatrix/RGBMatrix.c (hunk 1 above)
Comment (hierophect) [reply ports/stm/common-hal/rgbmatrix/RGBMatrix.c:39]: I'm sorry, can you rephrase? `stm_peripherals_timer_free` does undo never reset, is that your question? What do you mean by "remain"?

Comment (jepler) [reply ports/stm/common-hal/rgbmatrix/RGBMatrix.c:39]: CopyQ NkqthX

I'm still not fully familiar/comfortable with the transitions among these states; the naming is a bit different in each port and for each type of managed resource and I'm never entirely confident the semantics are the same.  (As a larger issue, it wouldn't hurt to document this for developers reference and to ensure consistency)

My question is whether the state transtion from "timer is in never_reset state" to "timer is in free state" by calling "timer_free" is a correct transition.  I think you answered yes.

Comment (hierophect) [reply ports/stm/common-hal/rgbmatrix/RGBMatrix.c:39]: Yes. I think the going concept here is that `never_reset` is actually a wholly separate option from `reserved` - it is possible to have an un-`reserved` pin that is not in `never_reset`. I suggested joining the two at one point, so that `never_reset` was strictly a subset of `reserved`, but I believe we decided against it (I don't fully remember the reason).

Conceptually, `reserved` is the idea that a pin/peripheral cannot be claimed twice within a single runtime. `never_reset` is the idea that resetting will not change the `reserved` status, or the pin attributes of that pin/peripheral (basically, immunity status to the `reset_all` function). To turn off `never_reset`, you typically call the single-reset function (the not reset-all one) which completely resets all aspects of the pin/peripheral. We have some `reset_ok` functions which specifically turn off `never_reset` and nothing else, but they're rarely implemented and even more rarely used.

I'd definitely be down to document this or remove any inconsistencies that you have found. I think we could also revisit whether the two ideas should be combined more concretely in some way - basically, is there ever a scenario where you want something to be in never_reset without it actually being claimed?
